### ISSUE TITLE
Feature Engineering: TDA features are global constants — replace with per-sample local topology

### ISSUE THREAD

**Issue by hedglinnolan** [hedglinnolan/tabular-ml-lab#138]
## Problem

The TDA section on Page 03 (`pages/03_Feature_Engineering.py`, ~lines 835-915) computes Vietoris-Rips persistent homology on the entire dataset treated as a single point cloud, derives summary statistics (entropy, amplitude, persistence-diagram point count per homology dimension), and then **tiles these globally-computed values across every row**:

```python
tda_features_array = np.array(tda_features_list).reshape(1, -1)
tda_df = pd.DataFrame(
    np.tile(tda_features_array, (n_samples, 1)),  # every row gets the same value
    columns=tda_feature_names,
    index=X_engineered.index
)
```

Result: the TDA columns have **zero variance across rows within any single train/test split**, so they contribute nothing to per-sample prediction. The UI presents them as "features" without flagging that they cannot drive R²/AUROC, which is misleading and risks reviewer pushback.

## Principle

Feature engineering should produce variables that augment the model's per-sample predictive power. Constant-across-rows descriptors are useful as *dataset-level metadata* (TRIPOD methodology, manuscript provenance) but are not features in the modeling sense and should not be presented as such.

## Proposal

Replace the global-TDA implementation with a **per-sample local topology** approach. Three candidate designs, ranked by defensibility:

### Option A — Local k-NN persistence (recommended)
For each row x_i, take its k nearest neighbors in feature space (typically k = 20–50), compute Vietoris-Rips persistence on the local point cloud, and derive scalar summaries (persistence entropy, max persistence, total persistence per homology dimension). Features vary per row by construction.

### Option B — Distance to topological landmarks
Compute a persistence diagram on a subsampled landmark set (~100 points), then for each row emit distance/similarity features (bottleneck distance to nearest landmark cycle, persistence-image density at the row's location, etc.). Features vary per row.

### Option C — Mapper-based cluster IDs
Run the Mapper algorithm on the dataset, assign each row to a Mapper node/cluster, and emit cluster-id features (one-hot or ordinal). Features vary per row.

Option A has the cleanest connection to standard persistent-homology theory and is most defensible to reviewers familiar with TDA literature.

## Implementation sketch

- New module `ml/topology_features.py` with `compute_local_tda_features(X, k=30, homology_dims=[0, 1], summary_metrics=['entropy', 'amplitude_landscape']) -> pd.DataFrame` returning one row per input sample.
- Replace the `np.tile` block on Page 03 with a call to the new function.
- Update the existing TDA expander text on Page 03: "Computes per-sample local topological features over each row's k-nearest neighbors. Cost scales as O(n · k³); subsampling rows during fit is recommended for n > 5000."
- Register an `InsightLedger` entry when local TDA is enabled, documenting feature additions and downstream-model considerations.
- `ml/narrative_engine.py`: add a brief Methods paragraph for local-TDA when the ledger entry is present.

## Acceptance criteria

- [ ] After enabling TDA, the resulting feature columns have non-zero variance across rows on every dataset.
- [ ] On a synthetic dataset with two distinct geometric structures (e.g., two concentric rings + noise), local TDA features split rows by structure with high purity (silhouette > 0.5).
- [ ] On NHANES yayhoo (`target=glucose`, n=21,849), local TDA features add a measurable, non-zero contribution to tree-based model R² above the baseline-without-TDA. Effect size doesn't need to be large but must be detectable.
- [ ] Page 03 UI documents the per-sample interpretation clearly so users understand what each feature represents.
- [ ] Unit test on a 200-row 2D point cloud with two clusters: H₀ entropy per row should differ meaningfully between cluster-core and cluster-boundary points.

## Caveats

- Computational cost is real. For n = 21k rows with k = 30, this is ~600k local persistence computations. Implementation needs caching, batching, and an explicit row-subsampling option (with the rest backfilled by nearest-neighbor inheritance).
- TDA features are notoriously hard to interpret. Coaching text should warn the user that downstream SHAP/permutation importance on local-TDA features will not yield clinically meaningful explanations.

## Related

- Discussed during demo prep on 2026-05-07. The broader principle motivating this issue is that **feature engineering operations should augment per-sample predictive power, not just produce dataset-level metadata**. If other Page 03 operations have similar global-vs-local issues, they should be audited under the same principle and filed as separate issues.
- Literature pointers: Carrière & Blumberg, "PersLay: A Neural Network Layer for Persistence Diagrams and New Graph Topological Signatures" (AISTATS 2020); Rieck et al., "Topological Machine Learning with Persistence Indicator Functions" (2019).

**Comment by hedglinnolan** [comment]
Status as of 2026-09-21: still open, nothing built. Keeping.

- The `np.tile` block is byte-for-byte unchanged; it moved from ~835–915 to `pages/03_Feature_Engineering.py:998-1002`. The TDA tab now spans lines 836–1027.
- The audit filed the same defect independently as `STATE-030` in `docs/turbotab/FINDINGS_LEDGER.md` ("landmine": the constants enter selection, training and importance rankings; the constant-column check runs on the audit frame before FE exists). Its minimum fix — drop constant TDA columns at creation, or record them as dataset-level metadata — is cheaper than this issue's per-sample design; the two are alternative resolutions of one defect.
- The only change since filing: PR #139 (Jul 18) builds the point cloud from training rows only under the sealed test set, and a code comment at `:931-933` says the columns are dataset-level constants. Nothing rendered on the page says so, and `ml/publication.py` still writes "Topological data analysis features" into the Methods as a named technique.
- New constraint: cohort branches replay feature-engineering steps from a recipe (`utils/replay.py`; `record` at `:90`, `replay_onto` at `:497`). Only the polynomial, interaction and ratio tabs record; TDA does not, so its columns surface as "unrecorded" on a cohort switch. Whatever replaces the tile must `_replay.record(...)` at the creation site.
- `ml/topology_features.py` does not exist; no test exercises page-03 TDA. Page 02 already computes dataset-level persistence diagrams and a Mapper graph as EDA (`ml/macro_shape.py`), which is the natural home for the global summaries if they become metadata.
- The NHANES acceptance criterion names a 21,849-row file that is not in the repo; it cannot run in CI as written.
